deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AutoPromoter.java

```diff
@@ -252,84 +437,82 @@ void processOnce(String envId) throws Exception {
         // How many candidates should we check
         int size = promoteBean.getQueue_size();
 
-        // If the service owner specify the cron expression, check if it's due for auto promote.
-        String schedule = promoteBean.getSchedule();
-        if (!StringUtils.isEmpty(schedule)) {
-            if (!autoDeployDue(currDeployBean, schedule)) {
-                return;
-            }
-            // For cron auto promote, we only choose the latest one.
-            size = Math.min(size, 1);
-        }
-
         // Promote build if preStage is BUILD
         if (promoteBean.getPred_stage().equals(Constants.BUILD_STAGE)) {
-            promoteBuild(currEnvBean, currDeployBean, size);
+            promoteBuild(currEnvBean, currDeployBean, size, promoteBean);
             return;
-        }
+        } else {
 
-        // Otherwise, promote from pred env deploys
-        promoteDeploy(currEnvBean, currDeployBean, size, promoteBean);
+            // If the service owner specify the cron expression, check if it's due for auto promote.
+            String schedule = promoteBean.getSchedule();
+            if (!StringUtils.isEmpty(schedule)) {
+                if (!autoDeployDue(currDeployBean, schedule)) {
+                    return;
+                }
+                // For cron auto promote, we only choose the latest one.
+                size = Math.min(size, 1);
+            }
+
+            // Otherwise, promote from pred env deploys
+            promoteDeploy(currEnvBean, currDeployBean, size, promoteBean);
+        }
     }
 
     boolean autoDeployDue(DeployBean deployBean, String cronExpressionString) {
         Date date = new Date();
+        return !date
+            .before(getScheduledCheckDueTime(deployBean.getStart_date(), cronExpressionString));
+    }
+
+    public Date getScheduledCheckDueTime(long start_date, String cronExpressionString) {
```

Comment:
For start_date, make naming consistent. Seems we prefer startDate. It would be nicer with startDateMs.